fnlb/lb/allgrouper.go

```diff
@@ -415,20 +370,26 @@ func (a *allGrouper) removeNode(w http.ResponseWriter, r *http.Request) {
 }
 
 func (a *allGrouper) listNodes(w http.ResponseWriter, r *http.Request) {
-	a.RLock()
-	nodes := make([]string, len(a.allNodes))
-	copy(nodes, a.allNodes)
-	a.RUnlock()
-
-	// TODO this isn't correct until at least one health check has hit all nodes (on start up).
-	// seems like not a huge deal, but here's a note anyway (every node will simply 'appear' healthy
-	// from this api even if we aren't routing to it [until first health check]).
-	out := make(map[string]string, len(nodes))
-	for _, n := range nodes {
-		if a.isDead(n) {
-			out[n] = "offline"
+
+	// get a list of node pointers
+	a.nodeLock.Lock()
+	run_list := make([]*nodeState, 0, len(a.nodeList))
+	for _, value := range a.nodeList {
+		run_list = append(run_list, value)
+	}
+	a.nodeLock.Unlock()
+
+	out := make(map[string]string, len(run_list))
+	for _, node := range run_list {
+
+		node.lock.Lock()
+		healthy := node.healthy
```

Comment:
would be nice to do this in the previous `nodeList` iteration if we get rid of granular locks.